trie-db/src/iterator.rs

```diff
@@ -315,6 +321,241 @@ impl<'a, 'cache, L: TrieLayout> TrieDBNodeIterator<'a, 'cache, L> {
+				},
+				Node::Branch(_, value) => match value {
+					Some(value) => value,
+					None => continue,
+				},
+				Node::NibbledBranch(partial, _, value) => {
+					prefix.append_partial(partial.right());
+					match value {
+						Some(value) => value,
+						None => continue,
+					}
+				},
+				_ => continue,
+			};
+
+			let (key_slice, maybe_extra_nibble) = prefix.as_prefix();
+			let key = key_slice.to_vec();
+			if let Some(extra_nibble) = maybe_extra_nibble {
+				return Some(Err(Box::new(TrieError::ValueAtIncompleteKey(key, extra_nibble))))
+			}
+
+			let value = match value {
+				Value::Node(hash) => match Self::fetch_value(db, &hash, (key_slice, None)) {
+					Ok(value) => value,
+					Err(err) => return Some(Err(err)),
+				},
+				Value::Inline(value) => value.to_vec(),
+			};
+
+			return Some(Ok((key, value)))
+		}
+		None
+	}
+
+	/// Fetches the next key.
+	///
+	/// Must be called with the same `db` as when the iterator was created.
+	pub fn next_key(&mut self, db: &TrieDB<L>) -> Option<TrieKeyItem<TrieHash<L>, CError<L>>> {
+		while let Some(raw_item) = self.next_raw_item(db) {
+			let (prefix, _, node) = match raw_item {
+				Ok(raw_item) => raw_item,
+				Err(err) => return Some(Err(err)),
+			};
+
+			let mut prefix = prefix.clone();
+			match node.node() {
+				Node::Leaf(partial, _) => {
+					prefix.append_partial(partial.right());
+				},
+				Node::Branch(_, value) =>
+					if value.is_none() {
+						continue
+					},
+				Node::NibbledBranch(partial, _, value) => {
+					prefix.append_partial(partial.right());
+					if value.is_none() {
+						continue
+					}
+				},
+				_ => continue,
+			};
+
+			let (key_slice, maybe_extra_nibble) = prefix.as_prefix();
+			let key = key_slice.to_vec();
+			if let Some(extra_nibble) = maybe_extra_nibble {
+				return Some(Err(Box::new(TrieError::ValueAtIncompleteKey(key, extra_nibble))))
+			}
+
+			return Some(Ok(key))
+		}
+		None
+	}
+}
+
+/// Iterator for going through all nodes in the trie in pre-order traversal order.
+pub struct TrieDBNodeIterator<'a, 'cache, L: TrieLayout> {
+	db: &'a TrieDB<'a, 'cache, L>,
+	raw_iter: TrieDBRawIterator<L>,
+}
+
+impl<'a, 'cache, L: TrieLayout> TrieDBNodeIterator<'a, 'cache, L> {
+	/// Create a new iterator.
+	pub fn new(db: &'a TrieDB<'a, 'cache, L>) -> Result<Self, TrieHash<L>, CError<L>> {
+		Ok(Self { raw_iter: TrieDBRawIterator::new(db)?, db })
+	}
+
+	/// Restore an iterator from a raw iterator.
+	pub fn from_raw(db: &'a TrieDB<'a, 'cache, L>, raw_iter: TrieDBRawIterator<L>) -> Self {
+		Self { db, raw_iter }
+	}
+
+	/// Convert the iterator to a raw iterator.
+	pub fn into_raw(self) -> TrieDBRawIterator<L> {
+		self.raw_iter
+	}
+
+	/// Fetch value by hash at a current node height
+	pub fn fetch_value(
+		&self,
+		key: &[u8],
+		prefix: Prefix,
+	) -> Result<DBValue, TrieHash<L>, CError<L>> {
+		TrieDBRawIterator::fetch_value(self.db, key, prefix)
+	}
+}
+
+impl<'a, 'cache, L: TrieLayout> TrieDBNodeIterator<'a, 'cache, L> {
```

Comment:
why you used a separated `impl` block for the functions that follows?

Reply:
That's how it was originally. (:

But good point, we can merge those.